### PR TITLE
[upgrade] Make system state object a dynamic field

### PR DESCRIPTION
**Pull request by lxfind**
In this PR, we change the system state object from wrapped to a dynamic field.
To do so we need to be able to derive the dynamic field ID. Moved that to a function.

### CODE REVIEW

**Review by amnn — COMMENTED** [review]
I think for now, the approach you have here, which uses a dynamic object field and caches the ID is the most straight forward.  In the non-object field case, you can derive the ID for the `Field` based on the ID of the parent (`0x5` in this case), the type of the field (`u64` in this case, for the protocol version), and the type of the inner type (this changes each time, but should be derivable from the protocol version).

It is more work, and requires exposing the way to calculate this ID to Rust (I think it's embedded into a native function at the moment), but otherwise, by depending on dynamic object fields here, I think we are signing on to supporting it everywere, ad infinitum. cc @tnowacki.

**Comment by emmazzz** [comment]
> by depending on dynamic object fields here, I think we are signing on to supporting it everywere, ad infinitum.

@amnn Why is that the case? We can always upgrade the framework code later and migrate the inner from a dynamic object field to a dynamic field right? The `inner_id` field will have to stay there, but that doesn't sound like a big deal.

**Comment by lxfind** [comment]
@amnn I have updated the PR to use dynamic field instead. Please take a look

**Review by amnn — CHANGES_REQUESTED** [review]
The change to `dynamic_field` looks good, thanks for changing that!  @emmazzz, yes, you're right in hindsight -- the only vestigial thing would be the `inner_id` field, so we could probably work around it in future, but seeing as it seems quite simple to use dynamic fields instead, it seems worth doing it now, and avoiding the confusion of a redundant field in future.

I just spotted that there are public function which return `SuiSystemStateInner` references -- this won't work unfortunately (when you update the framework, and introduce a new inner type for `SuiSystemState`, you won't be able to update the signatures of these functions.  Requesting changes for this.

Also, it is probably worth appending a version to the inner types (in Move) already, for our sanity, something like `SuiSystemStateInnerV0`.

File: crates/sui-framework/sources/governance/sui_system.move [hunk 1]
```diff
@@ -148,9 +145,8 @@ module sui::sui_system {
             // Use a hardcoded ID.
             id: object::sui_system_state(),
             version: protocol_version,
-            system_state,
         };
-        // dynamic_object_field::add(&mut self.id, self.version, system_state);
+        dynamic_field::add(&mut self.id, protocol_version, system_state);
```
Comment (mystenmark) [inline crates/sui-framework/sources/governance/sui_system.move:149]: i'm confused by this - if we advance the protocol version, how do we find this field afterward? I don't see any code yet that migrates the inner state from one version to the next.

Comment (lxfind) [reply crates/sui-framework/sources/governance/sui_system.move:149]: This is genesis, and I am making the initial version of system state the same as protocol version. 
We could also make them constant, if that's less confusing

Comment (lxfind) [reply crates/sui-framework/sources/governance/sui_system.move:149]: And upgrade logic hasn't been added yet

**Pull request merged**